### ISSUE TITLE
Encode Java types natively

### ISSUE THREAD

**Issue by olafurpg** [scalameta/scalameta#1703]
Currently, several Java types in SemanticDB have weird encodings

- primitive types: `scala/Int#`, `scala/Double#`, ...
- array types: `scala/Array#`
- wildcards: type ref with symbol `local_wildcard`

See spec https://github.com/scalameta/scalameta/blob/master/semanticdb/semanticdb3/semanticdb3.md#java-type

A core principle of Scalameta is to encode things for what they are so it might be an improvement to introduce native SemanticDB encodings for those types

- primitive types: `PrimitiveType(primitive: Primitive)`
- array types: `ArrayType(tpe: Type)`
- wildcard types: `WildcardType(tpe: Type, lower: Type, upper: Type)`

I'm not 100% convinced this is still worth it, so it needs further validation. The biggest problem I can think of is: what do we in Scala? We have primarily two options:

- Do the same in Java and Scala where possible. This has the benefit of consistency but the cost increasing the surface of the Scala type system.
- Keep the current Scala encoding unchanged. This has the benefit of having Scala signatures resemble closer to the Scala compiler at the cost of higher inconsistency between Java and Scala.

**Comment by xeno-by** [comment]
@olafurpg I looked into this ticket, and I agree that it seems like a significant change that's not necessarily a crystal clear win. While I'm not against experimentation (and I'm certainly fond of conceptual purity, which this ticket advocates for), the cost of the experiment seems pretty high. 

Concretely: I would prefer to keep the current Scala encoding unchanged (so that both Java and Scala represent their language features natively). Unfortunately, this means that unless we significantly improve the diff facilities in our expect tests, we're going to see a lot of spurious discrepancies (which is unacceptable, in my opinion). While I could certainly go ahead and introduce new types just for fun, spending several hours on diff facilities doesn't sound like fun (I did that way too much in Rsc, so at this point this kind of work isn't very enjoyable for me).

Therefore, since this ticket doesn't look like a blocker for our Rsc work, I'll be unassigning it from myself for the time being. Eventually, we'll start working on Java interop in Rsc, so it is likely that I won't be able to escape this fate, but for now I'd like to delay these fun times :)

To make sure we don't miss any blockers, I'll assign this ticket to you. If this ticket is also not on the critical path for your current work, please feel free to unassign - both from yourself and the 4.0.0-RC1 milestone.

**Comment by xeno-by** [comment]
You know, I could give it a try if we stop emitting SymbolInformation for Java compilation units in semanticdb-scalac (https://github.com/scalameta/scalameta/issues/1737). Then semanticdb-javac becomes the only source of truth, and we don't need to reconcile it with what Scalac thinks of Java types.

**Comment by xeno-by** [comment]
After sleeping on this and chatting with @olafurpg, I'm inclined to unassign this ticket from 4.0.0-RC1. 

This is a big change coming up at the 11th hour of the release, so it better be really important like the static inner class symbol format change. However, it just isn't, since it doesn't fix any blockers for our use cases, and it doesn't add any fundamental capabilities.

**Comment by olafurpg** [comment]
I agree with you @xeno-by that this change does not carry it's weight. I propose we close the ticket as "wontfix".

**Comment by xeno-by** [comment]
Let's keep this open for the time being. I suggest that when I'm back from vacation we revisit the Scalameta roadmap and then trim the tickets appropriately, not limiting ourselves to just this one.
